blog/2025-02-19-elastic-components.md

```diff
@@ -0,0 +1,520 @@
+controller maintains the controller.kubernetes.io/pod-deletion-cost label to ensure correct downsizing order.
+
+## Limitations of WorkloadSpread
+
+### Potential Risks of Webhook
+
+WorkloadSpread depends on Pod Mutating Webhook to function, which intercepts all Pod creation requests in the cluster.
+If the Webhook Pod (kruise-manager) experiences performance issues or failures, it may prevent new Pods from being
+created. Additionally, during large-scale scaling operations, Webhook can become a performance bottleneck.
+
+### Limitations of Acting on Pods
+
+While acting on Pods reduces business intrusion, it introduces limitations. For instance, CloneSet's gray release ratio
+cannot be controlled per Subset.
+
+## Case Study 1: Bandwidth Package Allocation in Large-Scale Load Testing
+
+A company needed to perform load testing before a major shopping festival. They developed a load-agent program to
+generate requests and used a CloneSet to manage agent replicas. To save costs, they purchased 10 shared bandwidth
+packages (each supporting 300 Pods) and aimed to dynamically allocate them to elastic agent replicas.
+
+They configured a WorkloadSpread with 11 Subsets: the first 10 Subsets had a capacity of 300 and patched Pod Annotations
+to bind specific bandwidth packages; the last Subset had no capacity and no bandwidth package, preventing extra
+bandwidth allocation if more than 3000 replicas were created.
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: WorkloadSpread
+metadata:
+  name: bandwidth-spread
+  namespace: loadtest
+spec:
+  targetRef:
+    apiVersion: apps.kruise.io/v1alpha1
+    kind: CloneSet
+    name: load-agent-XXXXX
+  subsets:
+    - name: bandwidthPackage-1
+      maxReplicas: 300
+      patch:
+        metadata:
+          annotations:
+            k8s.aliyun.com/eip-common-bandwidth-package-id: <id1>
+
+    - ...
+
+    - name: bandwidthPackage-10
+      maxReplicas: 300
+      patch:
+        metadata:
+          annotations:
+            k8s.aliyun.com/eip-common-bandwidth-package-id: <id10>
+    - name: no-eip
+```
+
+## Case Study 2: Compatibility for Scaling Managed K8S Cluster Services to Serverless Instances
+
+A company had a web service running on an IDC that needed to scale up due to business growth but could not
+expand the local data center. They chose to use virtual nodes to access cloud-based Serverless elastic compute power,
+forming a hybrid cloud. Their application used acceleration services like `fluid`, which were pre-deployed on nodes in
+the IDC but not available in the serverless subset. Therefore, they needed to inject a sidecar into cloud Pods to
+provide acceleration capabilities.
+
+To achieve this without modifying the existing Deployment's 8 replicas, they combined OpenKruise's SidecarSet and
+WorkloadSpread capabilities. They deployed a SidecarSet to inject an acceleration component into Pods labeled
```

Comment:
it is not necessary to mention sidecarset and Fluid does not use SidecarSet to inject sidecars